{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Feb. 7, 2017\n",
    "\n",
    "Problem: Low percentage of reads mapped to reference.\n",
    "\n",
    "Solution: Merge paired end reads before mutation-calling.\n",
    "\n",
    "Results:\n",
    "\n",
    "Sample, Initial%, Final%\n",
    "\n",
    "HA3-100, 67.9, 86.7\n",
    "\n",
    "HA3-300, 67.4, 85.3\n",
    "\n",
    "HA3-500, 87.9, 95.5\n",
    "\n",
    "HA3-780, 83.9, 93.2\n",
    "\n",
    "HA3-1000, 88.0, 98.7\n",
    "\n",
    "HE3-100, 83.1, 93.1\n",
    "\n",
    "HE3-300, 48.8, 71.5\n",
    "\n",
    "HE3-500, 90.1, 96.1\n",
    "\n",
    "HE3-780, 61.1, 79.6\n",
    "\n",
    "HE3-1000, NA, NA\n",
    "\n",
    "HR2-100, 87.7, 94.9\n",
    "\n",
    "HR2-300, 67.8, 85.6\n",
    "\n",
    "HR2-500, 86.6, 94.7\n",
    "\n",
    "HR2-780, 63.2, 82.9\n",
    "\n",
    "HR2-1000, NA, NA\n",
    "\n",
    "HS3-100, 89.5, 95.9\n",
    "\n",
    "HS3-300, 55.9, 75.4\n",
    "\n",
    "HS3-500, 84.9, 94.2\n",
    "\n",
    "HS3-780, 51.8, 73.9\n",
    "\n",
    "HS3-1000, 89.2, 98.6\n",
    "\n",
    "UA3-100, 88.0, 95.5\n",
    "\n",
    "UA3-300, 63.5, 81.5\n",
    "\n",
    "UA3-500, 78.3, 87.6\n",
    "\n",
    "UA3-780, 62.5, 82.3\n",
    "\n",
    "UA3-1000, 92.2, 95.8\n",
    "\n",
    "UE3-100, 88.8, 95.2\n",
    "\n",
    "UE3-300, 76.2, 90.8\n",
    "\n",
    "UE3-500, 81.7, 93.1\n",
    "\n",
    "UE3-780, 66.6, 84.7\n",
    "\n",
    "UE3-1000, NA, NA\n",
    "\n",
    "UR1-100, 88.0, 94.8\n",
    "\n",
    "UR1-300, 80.0, 91.3\n",
    "\n",
    "UR1-500, 76.5, 90.9\n",
    "\n",
    "UR1-780, 75.5, 89.5\n",
    "\n",
    "UR1-1000, 90.7, 94.9\n",
    "\n",
    "US1-100, 90.0, 95.6\n",
    "\n",
    "US1-300, 60.4, 79.0\n",
    "\n",
    "US1-500, 84.0, 94.3\n",
    "\n",
    "US1-780, 83.0, 91.9\n",
    "\n",
    "US1-1000, 88.5, 94.8\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Pipeline starts with pre-processed Illumina reads. \n",
    "#Pre-processing was done with sickle and scythe; for parameters, check with hilleskl/wkim.\n",
    "#This class is in /home/NETID/ymseah/Projects/Low_Mapping_in_breseq/bin/blastUnmapped.py\n",
    "\n",
    "import re, os, subprocess, time\n",
    "\n",
    "class run_software:    \n",
    "    def get_sample_name(self, fastq_file):\n",
    "        \"\"\"\n",
    "        Get sample names from fastq files that end in '1.fastq', '2.fastq', \n",
    "        or '_sing.fastq'.\n",
    "        \"\"\"\n",
    "        if re.search('[1-2].fastq|_[1-2].fastq|_R[1-2].fastq|_sing.fastq', fastq_file):\n",
    "            sample_name_end_index = re.search('[1-2].fastq|_[1-2].fastq|_R[1-2].fastq|_sing.fastq|__sing.fastq|_R_sing.fastq', fastq_file).start()\n",
    "            if re.search('/', fastq_file):\n",
    "                start_indices = [m.start() for m in re.finditer('/', fastq_file)]\n",
    "                start_indices.sort()\n",
    "                sample_name_start_index = start_indices[-1] + 1\n",
    "            else:\n",
    "                sample_name_start_index = 0\n",
    "            sample_name = fastq_file[sample_name_start_index:sample_name_end_index]\n",
    "            return sample_name\n",
    "        else:\n",
    "            print(fastq_file)\n",
    "            print('Input file name must end in \\'_1.fastq\\', \\'_2.fastq\\', \\'_R1.fastq\\', or \\'_R2.fastq\\'.')\n",
    "    \n",
    "    def get_all_sample_names(self, sample_dir):\n",
    "        sample_files = os.listdir(sample_dir)\n",
    "        sample_files.sort()\n",
    "        sample_names = []\n",
    "        for sample in sample_files:\n",
    "            sample_name = self.get_sample_name(sample)\n",
    "            if sample_name != None:\n",
    "                sample_names.append(sample_name)\n",
    "        return set(sample_names)\n",
    "    \n",
    "    def run_pear(self, fastq_f, fastq_r, output_dir, sample_name):\n",
    "        \"\"\"\n",
    "        PEAR usage: -f <fastq_f>, -r <fastq_r> -o <output_base_name>\n",
    "        \"\"\"\n",
    "        args = ['/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/bin/pear-0.9.10-bin-64/./pear-0.9.10-bin-64', '-f', fastq_f, '-r', fastq_r, '-o', output_dir+sample_name]\n",
    "        rp = subprocess.Popen(args)\n",
    "        #allow PEAR to be KeyboardInterrupted\n",
    "        try:\n",
    "            while rp.poll() is None:\n",
    "                time.sleep(0.1)\n",
    "        except KeyboardInterrupt:\n",
    "            rp.kill()\n",
    "            raise\n",
    "    \n",
    "    def batch_run_pear(self, sample_dir, output_dir):\n",
    "        sample_files = os.listdir(sample_dir)\n",
    "        sample_files.sort()\n",
    "        print('Getting sample names from FASTQ file names...')\n",
    "        all_names = self.get_all_sample_names(sample_dir)\n",
    "        fastq_f = []\n",
    "        fastq_r = []\n",
    "        for name in all_names:\n",
    "            if name + '1.fastq' in sample_files:\n",
    "                fastq_f.append(name + '1.fastq')\n",
    "            elif name + '_1.fastq' in sample_files:\n",
    "                fastq_f.append(name + '_1.fastq')\n",
    "            elif name + '_R1.fastq' in sample_files:\n",
    "                fastq_f.append(name + '_R1.fastq')\n",
    "        for name in all_names:\n",
    "            if name + '2.fastq' in sample_files:\n",
    "                fastq_r.append(name + '2.fastq')\n",
    "            elif name + '_2.fastq' in sample_files:\n",
    "                fastq_r.append(name + '_2.fastq')\n",
    "            elif name + '_R2.fastq' in sample_files:\n",
    "                fastq_r.append(name + '_R2.fastq')\n",
    "        \n",
    "        fastq_f.sort()\n",
    "        fastq_r.sort()\n",
    "        samples = list(all_names)\n",
    "        samples.sort()\n",
    "        \n",
    "        #check lengths of fastq_f and fastq_r\n",
    "        if len(fastq_f) != len(fastq_r):\n",
    "            print('Missing either forward/reverse read file(s)...')\n",
    "        else:\n",
    "            print('Equal number of forward and reverse read files.')\n",
    "            #check if all samples have read files\n",
    "            if len(samples) != len(fastq_f):\n",
    "                print('Missing read files for samples...')\n",
    "            else:\n",
    "                sample_counter = 0\n",
    "                while sample_counter < len(samples):\n",
    "                    if re.search(samples[sample_counter], fastq_f[sample_counter]) and re.search(samples[sample_counter], fastq_r[sample_counter]):\n",
    "                        print('Running PEAR on ' + samples[sample_counter])\n",
    "                        self.run_pear(sample_dir+fastq_f[sample_counter], sample_dir+fastq_r[sample_counter], output_dir, samples[sample_counter])\n",
    "                    sample_counter += 1    \n",
    "        \n",
    "    def get_all_pear_samples(self, pear_output_dir):\n",
    "        pear_files = os.listdir(pear_output_dir)\n",
    "        all_pear_samples = []\n",
    "        for file in pear_files:\n",
    "            filenamesplit = file.split('.')\n",
    "            if filenamesplit[1] == 'assembled' or filenamesplit[1] == 'unassembled' or filenamesplit[1] == 'discarded':\n",
    "                all_pear_samples.append(filenamesplit[0])\n",
    "            else:\n",
    "                all_pear_samples.append(filenamesplit[0] + '.' + filenamesplit[1])\n",
    "        return set(all_pear_samples)\n",
    "        \n",
    "    def run_breseq(self, input_dir, sample_name, output_dir, polymorphism_min = '5 ', ref_dir = '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/data/', ref_genome1 = 'dv.gbk', ref_genome2 = 'mp.gbk', ref_genome3 = 'megaplasma.gbk'):\n",
    "        \"\"\"\n",
    "        breseq usage: breseq -p -o . -r <ref_genome> --polymorphism-minimum-coverage-each-strand 5 <input_read_file>\n",
    "        \"\"\"\n",
    "        breseq_dir = output_dir + sample_name + '_breseq'\n",
    "        os.mkdir(breseq_dir)\n",
    "        ref1 = ref_dir + ref_genome1\n",
    "        ref2 = ref_dir + ref_genome2\n",
    "        ref3 = ref_dir + ref_genome3\n",
    "        reads1 = input_dir + sample_name + '.assembled.fastq'\n",
    "        reads2 = input_dir + sample_name + '.unassembled.forward.fastq'\n",
    "        reads3 = input_dir + sample_name + '.unassembled.reverse.fastq'\n",
    "        \n",
    "        args = ['breseq', '-p', '-o', breseq_dir, '-r', ref1, '-r', ref2, '-r',\n",
    "                ref3, '--polymorphism-minimum-coverage-each-strand', polymorphism_min, \n",
    "                reads1, reads2, reads3]\n",
    "        \n",
    "        rb = subprocess.Popen(args)\n",
    "        #allow breseq to be KeyboardInterrupted\n",
    "        try:\n",
    "            while rb.poll() is None:\n",
    "                time.sleep(0.1)\n",
    "        except KeyboardInterrupt:\n",
    "            rb.kill()\n",
    "            raise\n",
    "\n",
    "    def batch_run_breseq(self, pear_results_dir, breseq_output_dir):\n",
    "        all_samples = list(self.get_all_pear_samples(pear_results_dir))\n",
    "        for sample in all_samples:\n",
    "            self.run_breseq(pear_results_dir, sample, breseq_output_dir)\n",
    "    \n",
    "    def run_gdtools_compare(self, *samples, breseq_dir = '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/results/breseq_results/', \n",
    "                            ref_dir = '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/data/', \n",
    "                            ref_genome1 = 'dv.gbk', ref_genome2 = 'mp.gbk', ref_genome3 = 'megaplasma.gbk'):\n",
    "        \"\"\"\n",
    "        Usage: gdtools COMPARE [-o annotated.html] -r reference.gbk input.1.gd [input.2.gd ... ]\n",
    "        \"\"\"\n",
    "        ref1 = ref_dir + ref_genome1\n",
    "        ref2 = ref_dir + ref_genome2\n",
    "        ref3 = ref_dir + ref_genome3\n",
    "        ancestor = breseq_dir + 'sic_Ancestor_breseq/output/0.gd'\n",
    "        evo_line = samples[0]\n",
    "        evo_line = evo_line[4:]\n",
    "        evo_line_end_index = re.search('\\.|-|_', evo_line).start()\n",
    "        evo_line = evo_line[:evo_line_end_index]\n",
    "        output_file = breseq_dir + 'compare/' + evo_line + '.html'\n",
    "        args = ['gdtools', 'COMPARE', '-o', output_file, '-r', ref1, '-r', ref2,'-r', ref3, ancestor]\n",
    "\n",
    "        sample_counter = 0\n",
    "        while sample_counter < len(samples):\n",
    "            gd_file = 'output.gd'\n",
    "            gd_path = breseq_dir + samples[sample_counter] + '_breseq/output/'            \n",
    "            if re.search('-13', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-100.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath)\n",
    "            elif re.search('-14', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-100.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath)\n",
    "            elif re.search('-15', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-100.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath) \n",
    "            elif re.search('-16', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-100.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath) \n",
    "            elif re.search('-18', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-100.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath)\n",
    "            elif re.search('.45', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-300.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath)\n",
    "            elif re.search('-76', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-500.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath)\n",
    "            elif re.search('.118', samples[sample_counter]):\n",
    "                new_gd_filepath = gd_path + evo_line + '-780.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath)\n",
    "            else:\n",
    "                new_gd_filepath = gd_path + evo_line + '-1000.gd'\n",
    "                os.rename(gd_path + gd_file, new_gd_filepath)\n",
    "            args.append(new_gd_filepath)\n",
    "            sample_counter +=1\n",
    "        \n",
    "        rgc = subprocess.Popen(args)\n",
    "        #allow gdtools to be KeyboardInterrupted\n",
    "        try:\n",
    "            while rgc.poll() is None:\n",
    "                time.sleep(0.1)\n",
    "        except KeyboardInterrupt:\n",
    "            rgc.kill()\n",
    "            raise\n",
    "\n",
    "    def biopy_fastq_to_fasta(self, fastq_file, fasta_file):\n",
    "        \"\"\"Uses BioPython SeqIO module to convert fastq to fasta file\"\"\"\n",
    "        SeqIO.convert(fastq_file, \"fastq\", fasta_file, \"fasta\")\n",
    "        \n",
    "    def batch_biopy_fastq_to_fasta(self, input_dir = '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/results/breseq_results/'):\n",
    "        fastq_filepath = input_dir #+ sample_name + '_breseq/data/'\n",
    "        files = os.listdir(fastq_filepath)\n",
    "        fastq_filenames = []\n",
    "        fasta_filenames = []\n",
    "        for each in files:\n",
    "            if re.search('.unmatched.fastq', each):\n",
    "                fastq_filenames.append(each)\n",
    "        for fastq in fastq_filenames:\n",
    "            self.biopy_fastq_to_fasta(fastq_filepath + fastq, fastq_filepath + fastq[:-5] + 'fasta')\n",
    "            fasta_filenames.append(fastq[:-5] + 'fasta')\n",
    "        return fasta_filenames\n",
    "    \n",
    "    def run_qblast_fasta(self, fasta_file):\n",
    "        \"\"\"Iterate through fasta file to individually BLAST each read\"\"\"\n",
    "        for read in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "            result_handle = NCBIWWW.qblast(\"blastn\", \"nt\", read.format(\"fasta\"), hitlist_size=3)\n",
    "            with open(\"my_blast.xml\", \"a\") as blast_results:\n",
    "                blast_results.write(result_handle.read())\n",
    "\n",
    "    def run_blastn_remote(self, input_dir, fasta_file, output_dir):\n",
    "        \"\"\"Uses BLAST+ command line to search NCBI servers\n",
    "        blastn usage: ./blastn -db nt -query <input_file> -out <output_file> -remote\n",
    "        \"\"\"\n",
    "        blast_out = output_dir + fasta_file + 'BLASTout.txt'\n",
    "        args = ['/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/bin/ncbi-blast-2.5.0+/bin/./blastn',\n",
    "                '-db', 'nt', '-query', input_dir + fasta_file, '-out', blast_out, '-remote']\n",
    "        rbn = subprocess.Popen(args)\n",
    "        #allow blastn to be KeyboardInterrupted\n",
    "        try:\n",
    "            while rbn.poll() is None:\n",
    "                time.sleep(0.1)\n",
    "        except KeyboardInterrupt:\n",
    "            rbn.kill()\n",
    "            raise\n",
    "    \n",
    "    def batch_run_blastn_remote(self, list_of_files, input_dir = '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/results/blast_results/input', \n",
    "                                blast_out_dir = '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/results/blast_results/'):\n",
    "        for fasta in fasta_filenames:\n",
    "            self.run_blastn_remote(input_dir, fasta, blast_out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#This is found in /home/NETID/ymseah/Projects/Low_Mapping_in_breseq/bin/runPearBreseqBlast.py\n",
    "\n",
    "run_object = run_software()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Merging Reads\n",
    "\n",
    "Software: PEAR 0.9.10-bin-64\n",
    "\n",
    "Citation: Zhang, J. et al. PEAR: a fast and accurate Illumina Paired-End reAd mergeR. 2014. Bioinformatics 30(5):614-620. doi:10.1093/bioinformatics/btt593\n",
    "\n",
    "Parameters: default"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "run_object.batch_run_pear('/opt/data/wkim-data/FilteredFastQC/', '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/results/20161230/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Read Mapping & Mutation Calling\n",
    "\n",
    "Software: breseq 0.27.1\n",
    "\n",
    "Citation: Deatherage, DE., and Barrick, JE. Identification of mutations in laboratory evolved microbes from next-generation sequencing data using breseq. 2014. Methods Mol Biol. 1151:165-188. doi:10.1007/978-1-4939-0554-6_12\n",
    "\n",
    "Parameters: polymorphism mode (-p) default, except --polymorphism-minimum-coverage-each-strand 5 \n",
    "\n",
    "Notes: \n",
    "1. Input files are 3 PEAR output files: .assembled.fastq, .unassembled.forward.fastq, .unassembled.reverse.fastq\n",
    "2. Output directory has been moved to /opt/data/wkim-rsrch/breseq_results/."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "run_object.batch_run_breseq('/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/results/20161230/', '/home/NETID/ymseah/Projects/Low_Mapping_in_breseq/results/breseq_results/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Comparing Mutations Across Generations & Evolution Lines\n",
    "\n",
    "Software: breseq 0.27.1\n",
    "\n",
    "Notes:\n",
    "1. Sample script below is not for batch processing, but to rename input files to reflect generation/lines, before comparing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#This line is in /home/NETID/ymseah/Projects/Low_Mapping_in_breseq/bin/runGDT.py\n",
    "\n",
    "run_object.run_gdtools_compare('sic_UE3-14', 'sic_UE3.45', 'sic_UE3-76', 'sic_UE3.118')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Results\n",
    "\n",
    "wkim-ubuntu.css.uwb.edu/BreseqOutput/#/index\n",
    "\n",
    "Links to \"new result\" or \"NEWcompare...\" are from this analyses, i.e., merged reads prior to breseq."
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
